bip-0352.mediawiki

```diff
@@ -0,0 +1,457 @@
+* The sender should sign with one of the sighash flags ''ALL, SINGLE, NONE'' (''ANYONECANPAY'' is unsafe). It is strongly recommended implementations only use ''SIGHASH_ALL'' for silent payments<ref name="why_sighash_all">'''Why recommend ''SIGHASH_ALL''?''' Since the output address for the receiver is derived from from the sum of the ''[[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]]'' public keys, the inputs must not change once the sender has signed the transaction. If the inputs are allowed to change after the fact, the receiver will not be able to calculate the shared secret needed to find and spend the output. It is currently an open question on how a future version of silent payments could be made to work with new sighash flags such as ''SIGHASH_GROUP'' and ''SIGHASH_ANYPREVOUT''.</ref>
+* Inputs used to derive the shared secret are from the ''[[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]]'' list
+
+=== Scanning transactions ===
+
+A transaction is a ''Silent Payments v0'' transaction and MUST be scanned if and only if all of the following are true:
+
+* The transaction contains at least one BIP341 taproot output
+* The transaction has at least one input from the ''[[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]]'' list
+* The transaction does not contain a new, undefined output type (e.g. SegWit versions > 1)<ref name="skip_txs_with_unknown_prevouts">'''Why skip transactions that spend unknown output scripts?''' Skipping transactions that spend unknown output scripts allows us to have a clean upgrade path for Silent Payments by avoiding the need to scan the same transaction multiple times with different rule sets. If a fancy new output type is added in the future and Silent Payments v1 is released with support, we would want to avoid having to first scan the transaction with the silent payment v0 rules and then again with the silent payment v1 rules.</ref>
+* The transaction ''nVersion'' is < 4
+
+Otherwise, skip the transaction. This is to ensure forward compatibility with future versions of silent payments without requiring future versions to scan a transaction multiple times with different rule sets.
+
+=== Address encoding ===
+
+A silent payment address is constructed in the following manner:
+
+* Let ''B<sub>scan</sub>, b<sub>scan</sub> = Receiver's scan public key and corresponding private key''
+* Let ''B<sub>spend</sub>, b<sub>spend</sub> = Receiver's spend public key and corresponding private key''
+* Let ''B<sub>m</sub> = B<sub>spend</sub> + m·G'', where ''m'' an optional integer tweak for labeling
```

Comment:
Thanks for the review @Sosthene00 ! I think you're right that we need to specify this. Defining the label tweak as a 256 bit scalar seems the best choice, since this label `m` ends up as part of the private key when labels are used.

The reason for specifying it as an incremental integer is to easily recover your wallet from backup without any extra data about how many labels you used or derived them. The idea is you would pick some very large number `M` when scanning and check all integer values under that. But ultimately, it is up to the receiving client how to implement and I can imagine use cases where the label is an integer with some other meaning (like a product number or invoice number). I think maybe the best approach is like you said: specify that the label integer must be a 256-bit number, and we recommend making it an incremental scheme to enable easy backups.

Regarding the BIP32 idea, I'm unsure if it's possible to easily check if a public key exists in a given derivation path, without traversing it, right? So you'd still need a datastore to do the efficient "subtract and lookup" method when scanning.

Reply:
Yes I was making the assumption that the receiver maintains some kind of labels database to check against incoming candidate transactions. What you mean is that if we use a simple incremental scheme for labels we don't even need a database to find out outputs that belong to us (beside easy backup)? You still need some kind of data storage to match labels to some identity and/or context though, but even if you lose it you won't lose any of your money so that's still something I guess.